standard/expressions.md

```diff
@@ -6610,7 +6610,7 @@ constant_expression
     ;
 ```
 
-A constant expression may be either a value type or a reference type. If a constant expression is a value type, it must be one of the following types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type. If a constant expression is a reference type, it must be the `string` type, a default value expression ([§12.8.20](expressions.md#12820-default-value-expressions)) for some reference type, or the value of the expression must be `null`.
+A constant expression may have either a value type or a reference type. If a constant expression has a value type, it shall be one of the following types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type. If a constant expression has a reference type, it shall be the `string` type, a default value expression ([§12.8.20](expressions.md#12820-default-value-expressions)) for some reference type, or the value of the expression shall be `null`.
```

Comment:
This one feels a little odd to me. I think I'd prefer "The type of a constant expression may be a value type or a reference type. If the type is a value type [...]"

Reply:
Good suggestion for both the 2nd & 3rd sentences.

Reply:
wordsmithing as a suggestion:

```suggestion
The type of a constant expression may be either a value type or a reference type. If a constant expression has a value type, it shall be one of the following types: `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool,` or any enumeration type. If a constant expression has a reference type, it shall be the `string` type, a default value expression ([§12.8.20](expressions.md#12820-default-value-expressions)) for some reference type, or the value of the expression shall be `null`.
```

Reply:
The original text hails from the time when there was a null type, so “A constant expression may have either a value or a reference type…” made sense, but in modern C# `null` has no type… Also there is not really any need to call out value & reference types *per se* as a list of the valid types is provided.

This leads me to:

```suggestion
A constant expression shall either have the value `null` or one of the following types:
- `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, `bool`, `string`;
- an enumeration type; or
- a default value expression ([§12.8.20](expressions.md#12820-default-value-expressions)) for a reference type.
```

Reply:
I think the revised text misses that a constant expression may have no type, i.e. when its value is `null`.